process/elasticproc/logsevents/serializeDelegators.go

```diff
@@ -0,0 +1,152 @@
+		return nil, nil, errMarshal
+	}
+
+	meta := []byte(fmt.Sprintf(`{ "update" : { "_index":"%s", "_id" : "%s" } }%s`, index, converters.JsonEscape(id), "\n"))
+	if delegator.UnDelegateInfo != nil {
+		serializedData, err := prepareSerializedDataForUnDelegate(delegator, delegatorSerialized)
+		return meta, serializedData, err
+	}
+	if delegator.WithdrawFundIDs != nil {
+		serializedData, err := prepareSerializedDataForWithdrawal(delegator, delegatorSerialized)
+		return meta, serializedData, err
+	}
+
+	return meta, prepareSerializedDataForDelegator(delegatorSerialized), nil
+}
+
+func prepareSerializedDataForDelegator(delegatorSerialized []byte) []byte {
+	codeToExecute := `
+		if ('create' == ctx.op) {
+			ctx._source = params.delegator
+		} else {
+			ctx._source.activeStake = params.delegator.activeStake;
+			ctx._source.activeStakeNum = params.delegator.activeStakeNum;
+			ctx._source.timestamp = params.delegator.timestamp;
+		}
+`
+	serializedDataStr := fmt.Sprintf(`{"scripted_upsert": true, "script": {`+
+		`"source": "%s",`+
+		`"lang": "painless",`+
+		`"params": { "delegator": %s }},`+
+		`"upsert": {}}`,
+		converters.FormatPainlessSource(codeToExecute), string(delegatorSerialized),
+	)
+
+	return []byte(serializedDataStr)
+}
+
+func prepareSerializedDataForUnDelegate(delegator *data.Delegator, delegatorSerialized []byte) ([]byte, error) {
+	unDelegateInfoSerialized, err := json.Marshal(delegator.UnDelegateInfo)
+	if err != nil {
+		return nil, err
+	}
+
+	codeToExecute := `
+		if ('create' == ctx.op) {
+			ctx._source = params.delegator
+		} else {
+			if (!ctx._source.containsKey('unDelegateInfo')) {
+				ctx._source.unDelegateInfo = [params.unDelegate];
+			} else {
+				ctx._source.unDelegateInfo.add(params.unDelegate);
+			}
+
+			ctx._source.activeStake = params.delegator.activeStake;
+			ctx._source.activeStakeNum = params.delegator.activeStakeNum;
+			ctx._source.timestamp = params.delegator.timestamp;
+		}
+`
+	serializedDataStr := fmt.Sprintf(`{"scripted_upsert": true, "script": {`+
+		`"source": "%s",`+
+		`"lang": "painless",`+
+		`"params": { "delegator": %s, "unDelegate": %s }},`+
+		`"upsert": {}}`,
+		converters.FormatPainlessSource(codeToExecute), string(delegatorSerialized), string(unDelegateInfoSerialized),
+	)
+
+	return []byte(serializedDataStr), nil
+}
+
+func prepareSerializedDataForWithdrawal(delegator *data.Delegator, delegatorSerialized []byte) ([]byte, error) {
+	withdrawFundIDsSerialized, err := json.Marshal(delegator.WithdrawFundIDs)
+	if err != nil {
+		return nil, err
+	}
+
+	codeToExecute := `
+		if ('create' == ctx.op) {
+			ctx._source = params.delegator
+		} else {
+			if (!ctx._source.containsKey('unDelegateInfo')) {
+				return
+			} else {
+				for (int i = 0; i < ctx._source.unDelegateInfo.length; i++) {
+				  for (int j = 0; j < params.withdrawIds.length; j++) {
+				    if 	(ctx._source.unDelegateInfo[i].id == params.withdrawIds[j]) {
+						ctx._source.unDelegateInfo.remove(i);
```

Comment:
besides the double for, I don't think it's fine to remove an element from an array that you are iterating into. reference: https://www.java67.com/2018/12/how-to-remove-objects-or-elements-while-iterating-Arraylist-java.html

Reply:
refactored.